aexpect/remote.py

```diff
@@ -718,8 +720,142 @@ def scp_between_remotes(src, dst, port, s_passwd, d_passwd, s_name, d_name,
+    :param log_function: If specified, log all output using this function
+    :param timeout: The time duration (in seconds) to wait for the transfer
+                    to complete.
+    :param interface: The interface the neighbours attach to (only use when using
+                      ipv6 linklocal address).
+    :param directory: True to copy recursively if the directory to rsync
+    :raise: Whatever remote_rsync() raises
+    """
+    if limit:
+        limit = f"--bwlimit={limit}"
+
+    if host and host.lower().startswith("fe80"):
+        if not interface:
+            raise RsyncError("When using ipv6 linklocal address must assign",
+                             "the interface the neighbour attache")
+        host = f"{host}%{interface}"
+
+    command = "rsync"
+    if directory:
+        command = f"{command} -r"
+    command += (f" -avz -e 'ssh -p {port} -o UserKnownHostsFile=/dev/null "
+                f"-o StrictHostKeyChecking=no' {limit} "
+                f"{quote_path(local_path)} {username}@{host}:{shlex.quote(remote_path)}")
+    password_list = [password]
+    return remote_copy(command, password_list,
+                       log_filename, log_function, timeout)
+
+
+def rsync_from_remote(host, port, username, password, remote_path, local_path,
+                      limit="", log_filename=None, log_function=None,
+                      timeout=600, interface=None, directory=True):
+    """
+    Copy files from a remote host (guest) through rsync.
+
+    :param host: Hostname or IP address
+    :param port: Port
+    :param username: Username (if required)
+    :param password: Password (if required)
+    :param local_path: Path on the local machine where we are copying from
+    :param remote_path: Path on the remote machine where we are copying to
+    :param limit: Speed limit of file transfer.
+    :param log_filename: If specified, log all output to this file
+    :param log_function: If specified, log all output using this function
+    :param timeout: The time duration (in seconds) to wait for the transfer
+                    to complete.
+    :param interface: The interface the neighbours attach to (only use when
+                      using ipv6 linklocal address).
+    :param directory: True to copy recursively if the directory to rsync
+    :raise: Whatever remote_rsync() raises
+    """
+    if limit:
+        limit = f"--bwlimit={limit}"
+    if host and host.lower().startswith("fe80"):
+        if not interface:
+            raise RsyncError("When using ipv6 linklocal address must assign, ",
+                             "the interface the neighbour attache")
+        host = f"{host}%{interface}"
+
+    command = "rsync"
+    if directory:
+        command = f"{command} -r"
+    command += (f" -avz -e 'ssh -p {port} -o UserKnownHostsFile=/dev/null "
+                f"-o StrictHostKeyChecking=no' {limit} "
+                f"{username}@{host}:{quote_path(remote_path)} {shlex.quote(local_path)}")
```

Comment:
All of this is common to all three methods. How about splitting it into common part and then remote definition and share the common code that way?

I mean something like:

```python
def rsync_from_remote(...):
    command = _get_rsync_cmd(...)
    command += _get_rsync_host(...)
    command += "{shlex.quote(local_path)}"

def rsync_between_remotes(...):
    command = _get_rsync_cmd(...)
    command += _get_rsync_host(...)
    command += _get_rsync_host(...)
```

note it's not a strong request, I just think the less code one writes the easier is it to maintain it....

Reply:
In the future we might end up dropping scp entirely and just leaving rsync rendering such refactoring futile at present so I would keep the changes as additive as possible right now. Let's see after some more use of the remote module if it would make more sense to rather simplify it and maximize reusability.

Reply:
that seems reasonable